Minor refactor on top of existing FuncCall refactor.

wrzian opened the pull request:
Some changes relative to your PR here: https://github.com/typst/typst/pull/4435.
I left my main comment there to keep the discussion centralized: https://github.com/typst/typst/pull/4435#issuecomment-2210214435

This PR is mainly given as a suggestion, feel free to give feedback, reword comments, or let me know if you disagree with any changes.

wrzian committed:
Fix mutated method arg order. Break out early return in a new type

Coekjan commented on crates/typst/src/eval/call.rs:
```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
+) -> SourceResult<EvaluatedMethod> {
+    // Evaluate the method's target and overall arguments.
+    let (target, mut args) = if is_mutating_method(&method) {
+        // If this looks like a mutable method then we evaluate the arguments before the
+        // target because calling `target_expr.access(vm)` mutably borrows the vm, meaning
+        // we won't be able to evaluate args after due to the borrow checker.
```
```suggestion
        // If `method` looks like a mutating method, we evaluate the arguments first,
        // because `target_expr.access(vm)` mutably borrows the `vm`, so that we can't
        // evaluate the arguments after it.
```

Coekjan commented on crates/typst/src/eval/call.rs:
```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
+            target @ (Value::Array(_) | Value::Dict(_)) => {
+                let value = call_method_mut(target, &method, args, span);
+                let point = || Tracepoint::Call(Some(method.get().clone()));
+                return Ok(EvaluatedMethod::Special(value.trace(vm.world(), point, span)?));
```
Seems we need `rustfmt` here.

```suggestion
                return Ok(EvaluatedMethod::Special(value.trace(
                    vm.world(),
                    point,
                    span,
                )?));
```

wrzian replied:
It's one character over :sob:

wrzian replied:
Honestly I would rename the `EvaluatedMethod` enum just for this... maybe `MethodCall`? Thoughts?

Coekjan replied:
Ok to me

wrzian replied:
Ok, pushed this change.

Coekjan commented on crates/typst/src/eval/call.rs:
```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
+            target => (target.clone(), args),
+        }
+    } else {
+        // This is consistent with evaluating the callee first and args second as long as
+        // finding the method on the target is not affected by evaluating the args, which
+        // I believe is true
```
I think this comment can be removed, as the code statements below explain themselves already.

If we are not sure whether `target_expr.eval(vm)` affects the result of `args.eval(vm)`, I think we could comment the typst PR (not here) and I believe the maintainers will help us.

wrzian replied:
Not sure, I feel like I want to point out why the order is the way it is, maybe just shorten the comment to something like this?
> Function call order: we evaluate the method target before the arguments, but look up the method after.

Coekjan replied:
Or leave it here for maintainers to judge.

wrzian replied:
On second thought, I realize this is now commented above in the function's doc comment, so I'm just going to remove it.

Coekjan commented on crates/typst/src/eval/call.rs:
```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
         Value::Symbol(_) | Value::Func(_) | Value::Type(_) | Value::Module(_)
     ) {
-        Ok((target.field(&field).at(field.span())?, args))
+        // Certain value types have their own way to access method fields.
+        // Ex: `$arrow.r(v)$`, `func.with(fill: red)`
```
```suggestion
        // Certain value types may have their own ways to access method fields.
        // e.g. `$arrow.r(v)$`, `func.with(fill: red)`
```

Coekjan commented on crates/typst/src/eval/call.rs:
```diff
@@ -293,47 +281,92 @@ pub(crate) fn call_closure(
+        // Ex: `$arrow.r(v)$`, `func.with(fill: red)`
+        let value = target.field(&method).at(method.span())?;
+        Ok(EvaluatedMethod::Normal(value, args))
+        // Otherwise we couldn't find a mehtod to call for this type.
     } else {
```
```suggestion
    } else {
        // Otherwise we cannot find any methods to call for this type.
```

Coekjan reviewed (changes requested):
Great! Here are my thoughts:

wrzian committed:
Swap comment line order

Co-authored-by: Yip Coekjan <69834864+Coekjan@users.noreply.github.com>

wrzian committed:
Update some comment wording

Co-authored-by: Yip Coekjan <69834864+Coekjan@users.noreply.github.com>

wrzian reviewed (commented):
Applied some suggestions, but had some comments on others, lmk your thoughts!

Really appreciate the feedback :grin:

wrzian committed:
Renamed EvaluatedMethod to MethodCall

Honestly the main reason for this is to fix the formatter's complaint on line 307.

wrzian committed:
Remove extra comment about target/args call order.

See child PR discussion.

wrzian commented:
Updated with two commits, I'm good if you want to merge :+1:

Coekjan commented:
Let's merge it and continue discussing the refactor in the typst PR.
